File: download_libs.py
import urllib.request, os, sys

base = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                    'android', 'app', 'src', 'main', 'assets', 'www', 'lib')

print('base:', base, flush=True)

os.makedirs(base + '/fontawesome/css', exist_ok=True)
os.makedirs(base + '/fontawesome/webfonts', exist_ok=True)
os.makedirs(base + '/js', exist_ok=True)
print('dirs created', flush=True)

headers = {'User-Agent': 'Mozilla/5.0'}

def dl(url, path):
    try:
        req = urllib.request.Request(url, headers=headers)
        with urllib.request.urlopen(req, timeout=60) as r:
            data = r.read()
        with open(path, 'wb') as f:
            f.write(data)
        print('OK', os.path.basename(path), len(data), flush=True)
    except Exception as e:
        print('FAIL', url, str(e), flush=True)

FA_BASE = 'https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2'
dl(FA_BASE + '/css/all.min.css',               base + '/fontawesome/css/all.min.css')
dl(FA_BASE + '/webfonts/fa-solid-900.woff2',   base + '/fontawesome/webfonts/fa-solid-900.woff2')
dl(FA_BASE + '/webfonts/fa-regular-400.woff2', base + '/fontawesome/webfonts/fa-regular-400.woff2')
dl(FA_BASE + '/webfonts/fa-brands-400.woff2',  base + '/fontawesome/webfonts/fa-brands-400.woff2')

dl('https://cdnjs.cloudflare.com/ajax/libs/mammoth/1.4.21/mammoth.browser.min.js',
   base + '/js/mammoth.browser.min.js')
dl('https://cdnjs.cloudflare.com/ajax/libs/html2canvas/1.4.1/html2canvas.min.js',
   base + '/js/html2canvas.min.js')

print('ALL DONE', flush=True)

File: dev_tools/download_libs.js
const https = require('https');
const fs = require('fs');
const path = require('path');

const projectRoot = path.resolve(__dirname, '..');
const base = path.join(projectRoot, 'android', 'app', 'src', 'main', 'assets', 'www', 'lib');
console.log('base:', base);

fs.mkdirSync(path.join(base, 'fontawesome', 'css'), { recursive: true });
fs.mkdirSync(path.join(base, 'fontawesome', 'webfonts'), { recursive: true });
fs.mkdirSync(path.join(base, 'js'), { recursive: true });
console.log('dirs OK');

function dl(url, dest) {
  return new Promise((resolve, reject) => {
    const temp = dest + '.part';
    const file = fs.createWriteStream(temp);
    const request = https.get(url, { headers: { 'User-Agent': 'Mozilla/5.0' } }, res => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        file.close();
        res.resume();
        fs.unlink(temp, () => {});
        dl(res.headers.location, dest).then(resolve).catch(reject);
        return;
      }
      if (res.statusCode < 200 || res.statusCode >= 300) {
        file.close();
        res.resume();
        fs.unlink(temp, () => {});
        console.log('FAIL', url, 'HTTP ' + res.statusCode);
        resolve();
        return;
      }
      res.pipe(file);
      file.on('finish', () => {
        file.close();
        fs.renameSync(temp, dest);
        const size = fs.statSync(dest).size;
        console.log('OK', path.basename(dest), size);
        resolve();
      });
    }).on('error', err => {
      file.destroy();
      fs.unlink(temp, () => {});
      console.log('FAIL', url, err.message);
      resolve();
    });
    request.setTimeout(12000, () => request.destroy(new Error('timeout')));
  });
}

const FA = 'https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2';

(async () => {
  await dl(FA + '/css/all.min.css',               path.join(base, 'fontawesome', 'css', 'all.min.css'));
  await dl(FA + '/webfonts/fa-solid-900.woff2',   path.join(base, 'fontawesome', 'webfonts', 'fa-solid-900.woff2'));
  await dl(FA + '/webfonts/fa-regular-400.woff2', path.join(base, 'fontawesome', 'webfonts', 'fa-regular-400.woff2'));
  await dl(FA + '/webfonts/fa-brands-400.woff2',  path.join(base, 'fontawesome', 'webfonts', 'fa-brands-400.woff2'));
  await dl('https://cdnjs.cloudflare.com/ajax/libs/mammoth/1.4.21/mammoth.browser.min.js',
           path.join(base, 'js', 'mammoth.browser.min.js'));
  await dl('https://cdnjs.cloudflare.com/ajax/libs/html2canvas/1.4.1/html2canvas.min.js',
           path.join(base, 'js', 'html2canvas.min.js'));
  console.log('ALL DONE');
})();
